crates/core/tedge_api/src/workflow/mod.rs

```diff
@@ -57,13 +57,35 @@ pub enum OperationAction {
     /// ```
     MoveTo(StateName),
 
-    /// The built-in behavior is used
+    /// Implied built-in operation (for backward compatibility)
+    ///
+    /// - the operation name is derived from the workflow
+    /// - the step (trigger vs await) is derived from the command status (scheduled vs executing)
     ///
     /// ```toml
     /// action = "builtin"
+    /// on_exec = "<state>"
+    /// on_success = "<state>"
+    /// on_error = "<state>"
+    /// ```
+    BuiltIn(ExecHandlers, AwaitHandlers),
+
+    /// Trigger a built-in operation
+    ///
+    /// ```toml
+    /// action = "<builtin-operation-name>"
+    /// on_exec = "<state>"
+    /// ```
+    BuiltInAction(OperationName, ExecHandlers),
```

Comment:
Afterthought, I see that as over engineering.

The motivation was to be explicit:

- With `action = "builtin"` both the operation and the step are given by the context.
   - the operation is the operation of the workflow itself
   - the step is derived from the current state of the command (i.e scheduled or executing)
- With `action = "<builtin-operation-name>"` and `action = "await <builtin-operation-name>" both the operation and the step are explicit
   -  `action = "<builtin-operation-name>"` triggers the scheduled step of the given operation
   - `action = "await <builtin-operation-name>" triggers the execution step

But this raises other issues:
- It might then be confusing to be able to invoke an operation either directly (with `action = software_update`) or using a sub-workflow (with `operation = software_update`). These are meaningful and even useful differences, but this comes with complications notably to guide users.
- What if the user try to invoke the builtin behavior of operation A in a workflow for B? This could work but would requires extra work to handle the command payload, making even greater the confusion between builtin actions and sub operations.
- => In practice, __having the operation implied by the workflow is a good thing__,  simple and effective.

Hence, I'm considering to revert this change. i.e. to only keep the `action = "builtin"` case.

Another appealing alternative is the following:
- Keep the `action = "builtin"` as of now (honoring the user-provided handlers)
- Introduce two slight variants, to make the action clearer as it's not obvious of what is done by the `builtin` action depends on the state name (`scheduled` or `executing`).
- I propose: `action = "execute"` for the `scheduled` step (which can then be named differently).
- and `action = "await"` for the `executing` step.

Reply:
Here a new proposal https://github.com/thin-edge/thin-edge.io/issues/3014#issuecomment-2317687695.

I will address it in a different PR so we will be in position to compare and choose.